src/Workspaces/Core/Portable/Serialization/SolutionReplicationContext.cs

```diff
@@ -0,0 +1,32 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.Serialization
+{
+    internal readonly struct SolutionReplicationContext : IDisposable
+    {
+        private readonly ArrayBuilder<IDisposable> _resources;
+
+        private SolutionReplicationContext(ArrayBuilder<IDisposable> resources)
+            => _resources = resources;
+
+        public static SolutionReplicationContext Create()
+            => new(ArrayBuilder<IDisposable>.GetInstance());
+
+        public void AddResource(IDisposable resource)
+            => _resources.Add(resource);
+
+        public void Dispose()
```

Comment:
Do we need to call dispose explicitly when Scope is disposed?
[http://sourceroslyn.io/#Microsoft.CodeAnalysis.Remote.Workspaces/SolutionAssetStorage.Scope.cs,24](http://sourceroslyn.io/#Microsoft.CodeAnalysis.Remote.Workspaces/SolutionAssetStorage.Scope.cs,24)

Reply:
Oh, yeah. Good catch. I had that but then I stashed some changes and it disappeared. 🤦